### PR TITLE
CSHARP-2291: Allow unencoded subdelimiters in usernames and passwords

### PR DESCRIPTION
**Pull request by ksadoff**
https://jira.mongodb.org/browse/CSHARP-2291
https://evergreen.mongodb.com/version/5d1a7aa956234355ea3475d7

Ready for interim review.

There were several differences between our spec tests and the spec repo at the corresponding commit for this ticket, apart from the tests relevant to the ticket. After some discussion offline, when there were differences I defaulted to taking the spec repo version, since those changes were more recent.

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 1]
```diff
@@ -756,7 +756,7 @@ private void Parse()
             const string optionsPattern = @"(\?" + optionPattern + @"((&|;)" + optionPattern + ")*)?";
             const string pattern =
                 @"^(?<scheme>mongodb|mongodb\+srv)://" +
-                @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
+                @"((?<username>[^:@/]+)(:(?<password>[^:@/]*))?@)?" +
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:759]: I presume this change was necessary to handle [this test](https://github.com/mongodb/specifications/commit/4631ccd4f825fb1a3aba204510023f9b4d193a05#diff-84c3634193c53eae6131bd8cf7b788d6R196) that was made without a spec ticket?

**Comment by vincentkam** [comment]
Looks ready for team review to me!

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 1 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:759]: Yup!

**Comment by ksadoff** [comment]
Ready for final review.

**Review by vincentkam — APPROVED** [review]
LGTM 😸  Given that the other spec changes that the driver didn't have don't all seem to  have associated spec/csharp tickets, I'm fine with all the changes being done a single ticket

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/connection-string-spec.rst [hunk 2]
```diff
@@ -0,0 +1,386 @@
+.. role:: javascript(code)
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/connection-string-spec.rst]: No need to copy this file to our repo.

We only need to sync the `tests` directory itself.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/connection-string-spec.rst (hunk 2 above)
Comment (ksadoff) [reply tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/connection-string-spec.rst]: Done.

**Review by rstam — APPROVED** [review]

**Comment by ksadoff** [comment]
Rebased, merged, and pushed.

**Pull request closed without merging**